qiskit/circuit/quantumcircuit.py

```diff
@@ -1283,41 +1286,89 @@ def add_register(self, *regs: Union[Register, int, Sequence[Bit]]) -> None:
-                self._clbit_set.update(new_bits)
+
+                for idx, bit in enumerate(register):
+                    if bit in self._clbit_indices:
+                        self._clbit_indices[bit][1].append((register, idx))
+                    else:
+                        self._clbits.append(bit)
+                        self._clbit_indices[bit] = BitLocations(
+                            len(self._clbits) - 1, [(register, idx)]
+                        )
+
             elif isinstance(register, list):
                 self.add_bits(register)
             else:
                 raise CircuitError("expected a register")
 
     def add_bits(self, bits: Sequence[Bit]) -> None:
         """Add Bits to the circuit."""
-        duplicate_bits = set(self.qubits + self.clbits).intersection(bits)
+        duplicate_bits = set(self._qubit_indices).union(self._clbit_indices).intersection(bits)
         if duplicate_bits:
             raise CircuitError(f"Attempted to add bits found already in circuit: {duplicate_bits}")
 
         for bit in bits:
             if isinstance(bit, AncillaQubit):
                 self._ancillas.append(bit)
-
+                self._qubit_indices[bit] = BitLocations(len(self._ancillas) - 1, [])
             if isinstance(bit, Qubit):
                 self._qubits.append(bit)
-                self._qubit_set.add(bit)
+                self._qubit_indices[bit] = BitLocations(len(self._qubits) - 1, [])
             elif isinstance(bit, Clbit):
                 self._clbits.append(bit)
-                self._clbit_set.add(bit)
+                self._clbit_indices[bit] = BitLocations(len(self._clbits) - 1, [])
             else:
                 raise CircuitError(
                     "Expected an instance of Qubit, Clbit, or "
                     "AncillaQubit, but was passed {}".format(bit)
                 )
 
+    def find_bit(self, bit):
+        """Find locations in the circuit which can be used to reference a given Bit.
+
+        Args:
+            bit (Bit): The bit to locate.
+
+        Returns:
+            namedtuple(int, List[Tuple(Register, Int)]): A 2-tuple. The first element
+               (index) contains the index at which the Bit can be found (in either
+               `.qubits`, `.clbits`, depending on its type). The second element (registers)
+               is a list of Register-index pairs with an entry for each Register
+               in the circuit which contains the Bit (and the index in the
+               Register at which it can be found).
+
+        Note: The circuit index of an AncillaQubit will be its index
+            QuantumCircuit.qubits, not QuantumCircuit.ancilla.
+
+        Raises:
+            CircuitError: If the supplied Bit was of an unknown type.
+            CircuitError: If the supplied Bit could not be found on the circuit.
+        """
+
+        try:
+            if isinstance(bit, (Qubit, AncillaQubit)):
+                return self._qubit_indices[bit]
```

Comment:
`isinstance(bit, (Qubit, AncillaQubit))` is unnecessary since `AncillaQubit` is a subclass of `Qubit` - we can just do the first check.